feat(lock): allow for custom lock options

jacobheun opened the pull request:
* Allows users to create custom repo locks, such as to support the new s3 datastore
* ~~Allows users to have no repo locking~~

resolves ipfs/js-ipfs-repo#161

diasdavid commented:
@pgte @dignifiedquire can I get a review from you on this feature?

dignifiedquire commented on test/options-test.js:
```diff
@@ -28,6 +28,26 @@ describe('custom options tests', () => {
+
+  it('allows for no locker', () => {
+    const repo = new Repo(repoPath, {
+      locker: false
```
should also test for `undefined` options

dignifiedquire commented on test/node.js:
```diff
@@ -25,6 +41,16 @@ describe('IPFS Repo Tests onNode.js', () => {
     },
     init: true
+  }, {
+    name: 'custom locker',
```
There is no proper locking tests currently, so it would be good to test various locking scenarios in addition to the regular tests

dignifiedquire commented on src/index.js:
```diff
@@ -135,6 +134,65 @@ class IpfsRepo {
+   * @returns {Locker}
+   */
+  _getLocker () {
+    if (this.options.locker === false) {
```
This way of options is :-1: from my side. This forces `locker` to be of multiple types, so disabling the locker should check for `!= null`  and be if it is not set, rather than a boolean

Mr0grog replied:
I know multiple types for an option isn’t especially desirable, but it feels to me like just using `lock` for both custom and built-in types instead of adding a new option would be *much* nicer from the perspective of a user. Maybe we could move towards the type being a Locker object instead of a string? You could support built-ins with:

```js
// Expose the built-in collection as a frozen static var on Repo
new Repo({ lock: Repo.lockers.memory })
// OR
new Repo({ lock: require('ipfs-repo/lock-memory') })
```

…though we’d probably want to keep string support as a deprecated feature for a little while.

(Found my way here after reviewing ipfs/js-ipfs#1303; I hope I’m not butting in.)

dignifiedquire commented on src/index.js:
```diff
@@ -135,6 +134,65 @@ class IpfsRepo {
+   * @returns {void}
+   */
+  _openLock (path, callback) {
+    if (this._locker) {
```
also 👎 on being able to disable the locking, this is just asking for trouble

jacobheun replied:
I am fine with leaving locking required, the custom locker is the needed piece. It's fairly easy to subvert locking at the moment even if we don't explicitly allow it to be disabled. With custom datastore support there is a potential use case for users wanting to handle locking at the object level, rather than the repo level.

If we decide to leave it required, I can update the documentation when I add in the custom locker docs to encourage users to use locking appropriate for their use case and custom datastores, such as storing their lock on S3 if their datastore is there.

jacobheun commented:
@dignifiedquire @diasdavid

I've made a few updates after everyones feedback:
1) Documentation has been added to the Readme to help users see how to add a custom lock
2) I have removed the ability to disable locking, so this feature update can focus on the customization aspect only. This should give users the ability to add appropriate locking based on their datastore types.
3) I've added a suite to test that locks are being applied and that locking enforces one Node per Repo. 

I also updated the repo options to allow for overloading of the `lock` option as a string or Lock, instead of a separate property. As @Mr0grog mentioned, I also think this is more friendly for users. I noted that specifying `lock` as a string is *Deprecated*, but it is still currently supported to allow for backwards compatibility.

jacobheun commented:
@Mr0grog I updated things based on your comments, and I agree we should check for the `lock.close` method. I updated the code and added a test to verify it. If the assertion error is triggered, the internal lockfile reference won't be set, so `_closeLock` won't get called.
